src/client.rs

```diff
@@ -1,96 +1,126 @@
-use crate::{
-    extract::{self, Uri},
-    types::{Excludes, RequestMethod, Status},
-};
-use anyhow::{Context, Result};
+use crate::types::{Excludes, Response, Status, Uri};
+use anyhow::{anyhow, Context, Result};
 use check_if_email_exists::{check_email, CheckEmailInput};
+use derive_builder::Builder;
 use headers::{HeaderMap, HeaderValue};
 use hubcaps::{Credentials, Github};
-use indicatif::ProgressBar;
 use regex::{Regex, RegexSet};
 use reqwest::header;
 use std::net::IpAddr;
 use std::{collections::HashSet, time::Duration};
 use tokio::time::delay_for;
 use url::Url;
 
-/// A link checker using an API token for Github links
-/// otherwise a normal HTTP client.
-pub(crate) struct Checker<'a> {
+const DEFAULT_MAX_REDIRECTS: usize = 5;
+
+pub struct Client {
     reqwest_client: reqwest::Client,
     github: Option<Github>,
     includes: Option<RegexSet>,
     excludes: Excludes,
     scheme: Option<String>,
-    method: RequestMethod,
+    method: reqwest::Method,
     accepted: Option<HashSet<reqwest::StatusCode>>,
+}
+
+/// A link checker using an API token for Github links
+/// otherwise a normal HTTP client.
+#[derive(Builder, Debug)]
+#[builder(build_fn(skip))]
+#[builder(setter(into))]
+#[builder(name = "ClientBuilder")]
+pub struct ClientBuilderInternal {
+    github_token: Option<String>,
+    includes: Option<RegexSet>,
+    excludes: Excludes,
+    #[builder(default = "5")]
+    max_redirects: usize,
+    user_agent: String,
+    allow_insecure: bool,
+    scheme: Option<String>,
+    custom_headers: HeaderMap,
+    method: reqwest::Method,
+    accepted: Option<HashSet<http::StatusCode>>,
+    timeout: Option<Duration>,
     verbose: bool,
-    progress_bar: Option<&'a ProgressBar>,
 }
 
-impl<'a> Checker<'a> {
-    /// Creates a new link checker
-    // we should consider adding a config struct for this, so that the list
-    // of arguments is short
-    #[allow(clippy::too_many_arguments)]
-    pub fn try_new(
-        github_token: Option<String>,
-        includes: Option<RegexSet>,
-        excludes: Excludes,
-        max_redirects: usize,
-        user_agent: String,
-        allow_insecure: bool,
-        scheme: Option<String>,
-        custom_headers: HeaderMap,
-        method: RequestMethod,
-        accepted: Option<HashSet<http::StatusCode>>,
-        timeout: Option<Duration>,
-        verbose: bool,
-        progress_bar: Option<&'a ProgressBar>,
-    ) -> Result<Self> {
+impl ClientBuilder {
+    pub fn build(&mut self) -> Result<Client> {
```

Comment:
What if we make the `build` method consume `self` instead of taking a reference? E.g. see https://docs.rs/derive_builder/0.9.0/derive_builder/index.html#owned-aka-consuming

This would let us move things out of the builder into the resulting `Client`, improving the code in this function.

Also, if you come to think about it, should we expect for a builder to build more than one client? Maybe I'm missing some use case here, but I don't think the usual usage would involve building more than one client.

Reply:
Don't have any strong opinions on that. Guess we can experiment with consuming builders. Do you think we merge the current version in and then iterate on that later if the need arises? 😅

Reply:
Sounds good.